itensor/tagset.h

```diff
@@ -0,0 +1,228 @@
+//
+// Distributed under the ITensor Library License, Version 1.2
+//    (See accompanying LICENSE file.)
+//
+#ifndef __ITENSOR_TAGSET_H
+#define __ITENSOR_TAGSET_H
+#include "itensor/smallstring.h"
+
+namespace itensor {
+
+class TagSet;
+
+//
+// TagSet
+//
+
+class TagSet
+    {
+    public:
+    using tags_type = std::array<Tag,MAX_TAGS>;
+    private:
+    tags_type tags_;
+    size_t size_ = 0;
+    public:
+
+    TagSet() {};
+
+    //
+    // Convert a character array to a TagSet.
+    // Parses the array for ',' and splits up into tags, for example:
+    // "a,b,c" -> Tag("a"),Tag("b"),Tag("c")
+    TagSet(const char* t);
+
+    std::string
+    to_string()
+        {
+        std::string str = "";
+        for(size_t i = 0; i<size_-1; i++)
+            str = str+std::string(tags_[i])+",";
+        str = str+std::string(tags_[size_-1]);
+        return str;
+        }
+
+    const char*
+    c_str()
+        {
+        return this->to_string().c_str();
+        }
+
+    // 0-indexed access
+    Tag const&
+    operator[](int i) const
+        {
+        return tags_[i];
+        }
+
+    size_t
+    size() const {return size_;}
+
+    int
+    hasTag(const Tag& t) const;
+
+    void
+    addTag(const Tag& t);
+
+    void
+    addTags(const TagSet& ts);
+
+    void
+    removeTag(const Tag& t);
+
+    void
+    removeTags(const TagSet& ts);
+
+    };
+
+inline size_t
+size(TagSet const& t)
+    {
+    return t.size();
+    }
+
+inline bool
+operator==(TagSet const& t1, TagSet const& t2)
```

Comment:
Some compilers will complain that not all paths through this function will return a bool (even though logically it's not true). So a better design I think would be to remove the else branch and just put the for loop in the main function body. Then the first if will either return or fail, calling the "else" code anyway which is guaranteed to return.